### ISSUE TITLE
How to train PointPillar with multiple GPUs?

### ISSUE THREAD

**Issue by EthanZhangYi** [traveller59/second.pytorch#282]
When I train PointPillar with `second/configs/pointpillars/car/xyres_16.config` using 1 GPU, the best performance on KITTI val is `eval.kitti.official.Car.3d@0.70=[87.42, 76.53, 71.01]`. The performance reported in the paper is 77.98%. I think it is reproduced.

However, when I trained the model using 8 GPUs with `--multi_gpu=True`, I changed the `steps` to 37120 (1/8 of 296960), but the performance was very poor. I also tried to scale up the learning rate from 0.0002 to 0.0016, but the performance was still poor.

So how to train pointpillar with multiple GPUs?

**Comment by nydragon** [comment]
Bug in batch_idx in coordinates. One remedy is to add a line 

coors[:,0] -= coors[:,0].min()

after line 370 in voxelnet.py

**Comment by EthanZhangYi** [comment]
Hi, @nydragon 
Thanks for your reply. Could you please give a more detailed description of this bug?
After Line#370 in `voxelnet.py` is just code for evaluation. Just add `coors[:,0] -= coors[:,0].min()` for evaluation?
https://github.com/traveller59/second.pytorch/blob/3aba19c9688274f75ebb5e576f65cfe54773c021/second/pytorch/models/voxelnet.py#L339-L375

**Comment by nydragon** [comment]
sorry, my mistake, should be line 357

https://github.com/traveller59/second.pytorch/blob/3aba19c9688274f75ebb5e576f65cfe54773c021/second/pytorch/models/voxelnet.py#L357

**Comment by EthanZhangYi** [comment]
Thanks~ @nydragon 
So this line `coors[:,0] -= coors[:,0].min()` should be added in the `if` section after Line#357? Or it should be added outside the `if ` section?

**Comment by nydragon** [comment]
inside

**Comment by EthanZhangYi** [comment]
Thanks very much !!!

**Comment by HenryJunW** [comment]
@EthanZhangYi  How many epochs did you train? the result is at 160 epochs?

**Comment by EthanZhangYi** [comment]
@HenryJunW 
Single-GPU: total 296960 steps; 1856 steps per epoch; 160 epochs
Multi-GPU (8): total 37120 steps; 232 steps per epoch; 160 epochs

**Comment by HenryJunW** [comment]
Thanks. Did you get similar results for both a single GPU and 8 GPUs?

**Comment by EthanZhangYi** [comment]
Single-GPU better.
```
Car Result:
Single-GPU, batch_size=2: 86.29, 75.93, 69.83
Multi-GPU, batch_size=16: 84.78, 74.70, 68.25
```
For multi-gpu, we need to tune the hyper-parameters

**Comment by zhuMingXu** [comment]
Hi:
when I train second/configs/pointpillars/car/xyres_16.config (I don't change the file.)using 1 GPU.
I get the poor results?
Evaluation official
Car AP(Average Precision)@0.70, 0.70, 0.70:
bbox AP:50.13, 46.09, 41.36
bev  AP:49.07, 43.88, 40.43
3d   AP:45.29, 39.59, 34.53
aos  AP:0.13, 0.49, 0.63

train:3712  val:3769

I don't know why?

**Comment by zhuMingXu** [comment]
@EthanZhangYi

**Comment by EthanZhangYi** [comment]
Hi, @zhuMingXu 
For me, the performance can be reproduced without modifying the codes if a single GPU is used for training..

**Comment by Hub-Tian** [comment]
@EthanZhangYi How long does it take to train with multi-gpu? I use four gpus and reduce total steps to  74240. I expect the total training time to 1/4 compared with single gpu, however, it takes much longer time than one gpu......

**Comment by EthanZhangYi** [comment]
@Hub-Tian The speed is not satisfying.
with GTX1080-GPU,
1GPU, 296960 iterations, 22h
8GPU, 37120 iterations, 17h

**Comment by turboxin** [comment]
In my case the multi-gpu training is also slow, I thought it's because DataParallel, so I changed this to DataDistributedParallel, but this would consume much more gpu memory than 1gpu with same batchsize on one gpu. 
Anyone having the same problem?

**Comment by xmyqsh** [comment]
data loading，data augmentation，numpy的矩阵操作费时间，和点云数量成正比，可以试试nuscenes数据集的sweep10，更慢。我觉得data augmentation等矩阵操作可以放到gpu做，手撸cuda或者用pytorch写

**Comment by Shaoqing26** [comment]
Can u provide the training module for me.,Ia really need this, Thk.
